File: 977.squares-of-a-sorted-array.cpp
/*
 * @lc app=leetcode id=977 lang=cpp
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start
#include <algorithm>
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int32_t negative = 0;
        int32_t positive = nums.size() - 1;
        vector<int> answer;
        answer.reserve(nums.size());
        while (positive >= negative) {
            // both numbers are valid
            if (nums[positive] >= 0 && nums[negative] < 0) {
                if (nums[positive] >= -nums[negative]) {
                    answer.push_back(nums[positive] * nums[positive]);
                    positive--;
                } else {
                    answer.push_back(nums[negative] * nums[negative]);
                    negative++;
                }
            } else if (nums[positive] >= 0) { // only positive is valid
                answer.push_back(nums[positive] * nums[positive]);
                positive--;
            } else { // only negative is valid
                answer.push_back(nums[negative] * nums[negative]);
                negative++;
            }
        }
        std::reverse(answer.begin(), answer.end());
        return answer;
    }
};
// @lc code=end

File: 977.squares-of-a-sorted-array.rs
/*
 * @lc app=leetcode id=977 lang=rust
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut answer = vec![];
        let mut negative : i32 = 0;
        let mut positive : i32 = 0;
        while positive >= negative {
            // if both numbers are valid
            if nums[positive as usize] >= 0 && nums[negative as usize] < 0 {
                if (nums[positive as usize] >= -nums[negative as]) {
                    answer.push(nums[positive as usize] * nums[positive as usize]);
                    positive -= 1;
                } else {
                    answer.push(nums[negative as usize] * nums[negative as usize]);
                    negative += 1;
                }
            } else if (nums[positive as usize] >= 0) {
                answer.push(nums[positive as usize] * nums[positive as usize]);
                positive -= 1;
            } else {
                answer.push(nums[negative as usize] * nums[negative as usize]);
                negative += 1;
            }
        }
        answer.reverse();
        answer
    }
}
// @lc code=end

